sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1191 @@
+argument being `graph_state::modifiable` to reduce code verbosity on
+construction.
+
+.Graph State Diagram
+[source, mermaid]
+....
+graph LR
+    Modifiable -->|Finalize| Executable
+....
+
+==== Executable Graph Update
+
+A graph in the executable state can have each nodes inputs & outputs updated
+using the `command_graph::update()` method. This takes a graph in the
+modifiable state and updates the executable graph to use the node input &
+outputs of the modifiable graph, a technique called _Whole Graph Update_. The
+modifiable graph must have the same topology as the graph originally used to
+create the executable graphs, with the nodes added in the same order.
+
+==== Graph Member Functions
+
+Table 5. Constructor of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Constructor|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+command_graph(const property_list& propList = {});
+----
+|Creates a SYCL `command_graph` object in the modifiable state.
+Zero or more properties can be provided to the constructed SYCL `command_graph`
+via an instance of `property_list`.
+
+Preconditions:
+
+* This constructor is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `propList` - Optional parameter for passing properties. No `command_graph`
+  constructor properties are defined by this extension.
+
+|===
+
+Table 6. Member functions of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add(const property_list& propList = {});
+----
+|This creates an empty node which is associated to no task. Its intended use is
+either a connection point inside a graph between groups of nodes, and can
+significantly reduce the number of edges ( O(n) vs. O(n^2) ). Another use-case
+is building the structure of a graph first and adding tasks later.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The empty node which has been added to the graph.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+template<typename T>
+node add(T cgf, const property_list& propList = {});
+----
+|This function adds a command group function object to a graph. The function
+object statically contains a group of commands, of which a single command is
+executed at runtime. A function object can be a host task which is scheduled by
+the SYCL runtime, or a SYCL function for invoking kernels with all restrictions
+that apply as described in the core specification. The requisites of `cgf` will
```

Comment:
Are you intentionally disallowing the explicit memory operations here?  If so, why are they disallowed?

Reply:
No. Explicit memory operations are allowed as listed in the node section. It makes sense to explicitly list them here as well to avoid any confusion.

Reply:
This wording here still needs to be updated.  One option is to simply state that the command group function object may submit any command as defined by the core SYCL specification.

Reply:
Wording updated in https://github.com/reble/llvm/pull/117